rework the whole collect-filter-sort UI

AlicVB opened the pull request:
### Why ?
lot of issues / feature requests have been raised to point inconsistencies, issues, difficulty to use this key part of dt.
Even for experienced users, use of collect module is complex for lot of daily actions and tend to use the "quick & dirty" filter bar for that, which in turn tend to become more and more complex and big (see #6675 #10694)

this is also related to #10718 #10643 ...
### Idea
my idea here is to try to do the following :
First : no regression in actual features. Actual collect module is really powerful, and we have to keep that !
for the "easy to use" part :
- as few clicks as possible for usual features
- self-explanatory UI (as much as possible)
- avoid redundancy
for the "design" part :
- avoid cluttered UI
- let as much space as possible (in modules panels and in central view)

### basic idea
The main idea is to re-centralize everything in the collect module :
- the "collect" part
- the "recent collection" module
- the filtering bar
- the sort bar

Once done, we may also want to move the top bar icons (prefs, shortcuts, ...) to somewhere else (bottom bar ?) and gain 1 line for the images which would be really great imho !
Of course the main problem here is how to design the "new" collect module to avoid cluttering and make it easier to use...

### draft proposal
![collect2](https://user-images.githubusercontent.com/1818223/147661387-b89fc1eb-76a6-4d76-9c28-590f3dd2ecc1.png)
- each rule can be shown in 2 way : the "raw" entry which is the same as the current one *or* by default, more friendly widgets (in the example screenshot, I've reused the filter rating part...) One can switch between the 2 with the icon on the right
- current "tree/list" part can be shown for each rule with the expand button *and* using the completion popup of the "raw" entry
- in the "add rule" popup, the first lines allow to *append* presets to the existing rules
- the right icon of the last line show the list of the recently used collection (and replace the module)
- one can add several sort order if needed
- rule and sort blocks can be reordered by mouse if needed
- several fix / enhancement need to be done to the actual filters (rating and rejected, relative date-time, etc...)
- on preset saving, ask the user which rule/sort he want to save. So one can save "basic presets" or "filtering presets"
- for default rules, I would propose to have "folder -all- + rating -all-" so that would improve discoverability and we are 1:1 to the actual situation, with the "filter bar" content inside the collect module with as few clicks needed to change them

###  Question :
the main one : Do we need this ? (I'm convinced we do, but I'm not alone here !)

### about this PR
Be careful : the PR content is far from finish, only the basics is functional (but buggy)... No need to point bugs or whatever at the moment ! (there would be too many)
This is a huge task. Once we agree (or not) with the main things, I will accept gracefully if some of you want to help me, esp. for the "friendly widgets" part !

MStraeten commented:
hmm, i find the rating and color marks in the headline more comfortable since i use these to work on my current collection:
- blend in/out higher/lower rated images while editing (lighttable and darkroom: e.g. do the better first,...) 
- distinguish purpose of edits (filter selection by color marks to export subsets)
- ...

of course this can be done as a part of an boolean term to define a collection, but from a non technical view i'm using this to work inside a collection, not to define a collection - and it’s better to be able to use this even the side panels are blend out.
So i doubt moving this into collection gives a better user experience ...

If thats an optional enhancement for the collection and rating/color marks can be used as is in top line (bottom line s used to define those) i'm fine with that - having options is always better;)

AlicVB commented:
I'm really not convinced with the distinction between collection and filtering. All of this fall in the same query at the end. As a proof , we can see the actual discussions about what to put in the filter bar... potentially almost every "collect rule" is legitimate to appear there, depending on user workflow...
The main usage of this distinction I can see, is for presets, when you can have "base presets" and "filtering presets" that will append (and not put in place) to the actual rules. I've handled that in my proposal too (I've just edit to add it, as I forgot this point)

Now for the UI, my goal is to make the filtering bar useless by enhancing the collect module, but it's not a problem to keep it in parallel. Once the work done in the collect module, we'll have to decide what to do for the filtering bar : 
- keep it as it is (your preference iiuc)
- change it so it's directly related to the collect module
- change it completely for some "visual shortcuts" to the collect module (like we do with QAP)
- remove it (with all the drawbacks you mention)

TurboGit commented:
> but from a non technical view i'm using this to work inside a collection, not to define a collection - and it’s better to be able to use this even the side panels are blend out.

I have the same way of working.

TurboGit commented:
> * change it completely for some "visual shortcuts" to the collect module (like we do with QAP)

I suppose this would work for me, yes. Anyway let's see how all this bring to us...

quovadit commented:
I like the idea to merge collection and filtering technically, and then use the filtering bar as "visual shortcuts" for the collections-module.

But to be honest I don't see the advantage of your UI.
(maybe because it's not finished yet...).
Which filters should already work? What does the arrow down (v) do?
![image](https://user-images.githubusercontent.com/5677563/147678764-462bb898-fc34-40d5-bc4f-decfbc20fc36.png)


I did some mockups [here](https://github.com/darktable-org/darktable/issues/9664#issuecomment-890117538) and [here](https://github.com/darktable-org/darktable/issues/10643#issuecomment-999912297), which are much more tidied up in order to be "easy-to-use" in my opinion.

AlicVB commented:
@quovadit : thanks for the test ! the arrow are used to expand the rule in order to show the list/treeview part. But as said all this is currently WIP at a very early stage and they are not implemented atm. the list/treeview is also planned to be shown in the regular autocompletion popup of the entry.

I've already seen your screenshots, thanks for them.
The goal with my design here is : 
- allow more space for the "important info part" : either the "raw" entry either the "friendly" widgets
- each rule is clearly distinct from the other and "self-contained" : operator inside, listview/treeview inside, etc...

as said, not sure it's worth the pain to test the code actually, it's buggy, unfinished, miss most features, ...

quovadit commented:
@AlicVB : thanks for the explanation, expanding list/treeview sounds great.

After some deeper thinking, I really like your allocation of the elements in blocks.

I tried some cleanup of your draft:
![image](https://user-images.githubusercontent.com/5677563/147687500-fcb30563-c700-43e2-bd52-6e72dcc57bdb.png)

- doubled space between criteria-blocks
- remove headline 'collect rules'
- center the name of the criterion
- button 'new rule...' inside a new criteria-block
- possibility to add another sort-criteria with an empty sort row
- actions-block removed

I don't see the benefit of integrating the recently used collections, I would prefer to extend it to a more powerful 'recent' module as shown [here](https://github.com/darktable-org/darktable/issues/10643#issuecomment-999912297)

AlicVB commented:
Thanks for your ideas and mockup !



> * doubled space between criteria-blocks
> 
> * remove headline 'collect rules'

yep, I'll do it



> * center the name of the criterion

that's more challenging due to gtk, but it's on my todo too


> * button 'new rule...' inside a new criteria-block
> 
> * possibility to add another sort-criteria with an empty sort row

I'm not a great fan, imo it clutter the crucial parts of the module, where you want to see very quickly what is there or not



> I don't see the benefit of integrating the recently used collections, I would prefer to extend it to a more powerful 'recent' module as shown [here](https://github.com/darktable-org/darktable/issues/10643#issuecomment-999912297)

yes, I've seen your proposal. Why not, but that can be integrated without problems to my proposal, just some submenus to add.
I would prefer to integrate recent collections inside collect. That belong to the same things and the functionality doesn't need so much place and complexity imo

phweyland commented:
> I'm really not convinced with the distinction between collection and filtering

Technically that's the same query, you are right.
The difference is in the usage, in granularity and versality.
First one chooses the subject, then refines/works/cules: collect module is perfect for the first step, including its presets. 
Then one works on that subject: quick filter bar is perfect for quick / versatile work. 

Must be recalled here that the current collect presets behavior is not compatible with quick filtering (probably a good indice that collect module is not for that purpose). 

As @MStraeten, I quite prefer the quick filter bar to support work on the images. 

For me it isn't worth to sacrify the quick filter bar to save space. I prefer to improve the quick filters to make a better use of that space.

quovadit commented:
> For me it isn't worth to sacrify the quick filter bar to save space

I absolutely agree. It's important to have a quick filter, always visible, even in darkroom or if the left sidebar in lighttable is collapsed.

But nevertheless I think it's better to technically integrate the functionality to one place, and use the quick-filter as a visual shortcut.

elstoc commented:
I tend to go through a fairly standard process when I'm doing image culling in the lighttable, whereby I start with everything at one star and then gradually select or reject images by increasing to two or decreasing to zero stars. While I'm doing that I may change collection (folder) frequently, while staying at the same point in my editing process (for example always viewing one star images only). It would be nice to be able to do something similar with colour labels. I will then change my filter and repeat, again possibly in multiple different folders.

The important thing here is that I want to retain the ability to change "collection" while keeping my "quick filter" unchanged, i.e. that the two remain independent of one another and that, for example, resetting the collections module does not reset the quick filter. Similarly I don't want to make the initial construction of such a filtered collection more complex. I'm worried that incorporating the "quick filters" into the collections module might make this workflow more fiddly to implement.

quovadit commented:
> I want to retain the ability to change "collection" while keeping my "quick filter" unchanged

I understand your workflow,  but I don't see why this should not work with the new collect-filter.

It should be even simpler to quickly change folder while keeping rating and vice versa, if those two criteria are expanded in collection module.

elstoc commented:
I should add that I'm strongly in favour of refactoring the underlying code (afaict it's grown quite organically and it's not very pleasant to read/understand - I suspect it's fairly inefficient too) and I don't at all object to refactoring of the UI if it can be improved. If the workflow I've mentioned can be easily supported then that's great. I'll admit I haven't looked at the suggestions here in great detail but just wanted to state my use case to ensure it's considered.

An option (I'm not saying it's the best option) might be to allow the quick filter rules to be pinned to the top of the module (or always present) so that they're not affected by a module reset. So for example by default three rules could be shown (color: any, star: any, film roll: user selected).

AlicVB commented:
@phweyland @elstoc : I understand your concern, and I have the same workflow as you, except that what I mostly need in the filtering, apart from rating and colorlabels, is a filter by camera model : my wife and I have different cameras with different quality and I often need to browse only my photos (best quality) among all the images in a folder. That can be done also with a filter by extension as they have different extension...
All that to show that almost *all* rules are legitimate to be in the filter bar, depending on your workflow...

For now, I think it's better to concentrate to the collect module only which need *huge* work as @elstoc points, and let the filter bar apart (we can decide later what to do for it) The only point that will have to change is the sorting part. For that one, I propose to always link the sort combo of the filter bar to the first one in collect. That may please everyone for instance !

(tbh, currently, I have a special import script in lua that rates *my* images at 2, those of my wife to 1 and the other to 0... but that's a crappy workaround, not possible for everyone...)

phweyland commented:
> always link the sort combo of the filter bar to the first one in collect

Is that just because the first one is always present? Or do you have any other reason?
For the user point of view it's applied on top of all...

phweyland commented:
I want to add that one of the benefits of the quick filter bar exposes the filters. They are always visible.
Once included in one of the layers of collect module you lose this permanent contact,  which can source of mistake. 
I agree that collect module needs some love, but I don't see how the huge list shown above

quovadit commented:
> I don't see how the huge list shown above can be more friendly...

I agree, the long list must not get even longer.

So I would definitely split appending presets from adding rules.

I would even go further and  propose a separate edit-mode, where rules can be added, deleted, reordered and changed.

In view mode you of course change the values of the shown rules, and load (=replace) or append (=add) saved presets.

![image](https://user-images.githubusercontent.com/5677563/147738238-c56b76f6-5b8b-4066-bd92-0150e83563c6.png)

AlicVB commented:
OK OK... let's say it once for all : I've understand, and I will NOT remove the filter bar :)
My goal is not to oppose to any one or break workflow !

Anyway, the amount of work in collect only is huge, I will concentrate on that for instance. Just making things "working" is a challenge and a battle against gtk (as usual ;)

Nilvus commented:
> I like the idea to merge collection and filtering technically, and then use the filtering bar as "visual shortcuts" for the collections-module.
> 
> But to be honest I don't see the advantage of your UI. (maybe because it's not finished yet...). Which filters should already work? What does the arrow down (v) do? ![image](https://user-images.githubusercontent.com/5677563/147678764-462bb898-fc34-40d5-bc4f-decfbc20fc36.png)
> 
> I did some mockups [here](https://github.com/darktable-org/darktable/issues/9664#issuecomment-890117538) and [here](https://github.com/darktable-org/darktable/issues/10643#issuecomment-999912297), which are much more tidied up in order to be "easy-to-use" in my opinion.

I completely agree. @AlicVB: I agree with you too that this need to be improved and, as @quovadit, the idea to merge collection and filtering. Even if I have understand that this is just a draft, I don't like this first proposal. I don't understand it and find it worse than actual UI. And yes, what are those arrows. I think we need to discuss it and find a better way before going further (and avoid you to lost too much time).

AlicVB commented:
@Nilvus : happy new year ! I was wondering why you didn't say anything until now, although I know that you always love my drafts :P (hopefully you are here to finetune them ;))
As you say, it's a very first draft. Here's how work the current design (which can and will change) : 
1. each filter is self-contained => avoid the "click on the first rule, see the tree after the last rule"
2. each filter expose 2 sets of widgets : the "raw" entry (the one which was used before) and some more friendly widgets. The switch between raw/friendly is done with the icon under the close button (yes the icon is ugly :))
3. when typing on the raw entry, we show the filtered list of proposal using the standard entry autocompletion
4. the full list of proposals is optionally shown in a tree/list view which can be shown/hidden using the infamous arrows. In fact, showing the full list permanently is not needed for most of the rules, that's why I've done it under an expander (I've tried with a "standard expander, but that was ugly and very distractive imho)

I hope that I've been clear enough so that you at least understand the UI (not sure)

Nilvus commented:
> @Nilvus : happy new year ! I was wondering why you didn't say anything until now, although I know that you always love my drafts :P (hopefully you are here to finetune them ;)) As you say, it's a very first draft. Here's how work the current design (which can and will change) :
> 
>     1. each filter is self-contained => avoid the "click on the first rule, see the tree after the last rule"
> 
>     2. each filter expose 2 sets of widgets : the "raw" entry (the one which was used before) and some more friendly widgets. The switch between raw/friendly is done with the icon under the close button (yes the icon is ugly :))
> 
>     3. when typing on the raw entry, we show the filtered list of proposal using the standard entry autocompletion
> 
>     4. the full list of proposals is optionally shown in a tree/list view which can be shown/hidden using the infamous arrows. In fact, showing the full list permanently is not needed for most of the rules, that's why I've done it under an expander (I've tried with a "standard expander, but that was ugly and very distractive imho)
> 
> 
> I hope that I've been clear enough so that you at least understand the UI (not sure)

@AlicVB: happy new year. Quite few time actually. I will read that better later and try to think about some proposals. It could take time actually.

Just one thing: we need to keep actual list of collections regarding first filter. I like to see all my filmrolls/folder and click on them to see thumbnails.

AlicVB commented:
... and I've just pushed some commit to propose something a little more functional : tree/list view can be shown, autocompletion works (for list only)...
But as before, this is still full of bugs : switch between raw/friedly doesn't always works, sometimes the tree/list view is empty or not refreshed, autocompletion is very basic, ... and "friendly" widgets are done only for rating atm...

But it should give testers a better idea how that works...

AlicVB commented:
> I like to see all my filmrolls/folder and click on them to see thumbnails. On my side, if arrows were for that, they didn't work in that PR by now. And i prefer have a first expanded list like actually.

yes, that was not working (before my last commits) For the expanded state, my idea (not implemented) would be : 
1. store it's state per rule type to remember that for the next use.
2. define the default expand state to "on" or "off" depending on the rule type (useful for folders, tags, ... not so much for ratings, colorlabels, dates, even filmrolls)

elstoc commented:
Might be useful to be able to see and even manually adjust the SQL that the collections module generates, since right now, we are quite heavily restricted by the sort of (mostly string) comparisons the UI allows. For example, one can't easily do queries that only impact the time of date/time fields (see https://discuss.pixls.us/t/filter-by-time-in-collection/28757 for a discussion on this subject).

AlicVB commented:
@elstoc : date-time part clearly need some love. There's your example, but also relative date-time, ... It's on my todo, but in a second part, once the global design will be stabilized...
Now, you're right to say that this won't cover all the need, it's simply impossible. Thus having a sort of "full raw" mode when user can write himself the query might be a solution... with some drawbacks, esp. about the handling of "bad" queries which may potentially void the whole database... and yes, we will only be able to switch from "classic" interface to "raw" and not the contrary.

EC1000 commented:
Something I would personnally find a nice ergonomic addition would be to be able write the filtering attribute instead of just choosing it in the list. Since the list is relatively long, it always take a few seconds to find the position in the list of the one I am looking for. For example, enter "col..." + Enter and it would select "Color label" attribute.

AlicVB commented:
Some update on this : the new version should almost "work" now even if not finished or complete and certainly not bug-free :)
- "friendly widgets" are done for rating and aspect ratio only for instance
- sort widget in collect and in filter bar are linked together
- list view/treeview should be ok, apart the update when a previous rule change
- completion suggestion should be implemented too. Note that currently I didn't manage to show it as a tree. Seems to be a gtk bug...
- appending presets should be ok. (I've done 2 presets for example purpose only)
- I've removed the old presets which should be handled by enhancing the widgets (ratios) or the query itself (date-time) Once everything in place we'll have to rethink that...
- expanded state and raw/friendly state are stored by rule property, so you if you prefer having tags or folders always expanded, that's possible. Currently default states are "not expanded" and "show friendly widgets" but that can be changed easily by setting some default values in `darktableconfig.xml.in`
- secondary sorts is not implemented
- reordering rules is not implemented
- there's an issue with collect lib sizing

Thanks for your tests and feedbacks

phweyland commented:
I like the button `revert to a previous set of rules` (avoids to open another module). It could be put in the module header or in a section of presets.
The expansion button could be included in the header of the sub-modules to save space.
It should be also expanded by default.

Appending a preset is great as a function, but the long list of rules is not appealing.
It would more logic to append presets from the presets menu (probably having a specific section for them). Currently they appear in presets but act as normal presets which not (mandatorily) wanted.
The long list could be also reduced by a configuration as for `image information` to keep only the required ones.

If I want to change the main rule from folder to tag I have to add a new rule and remove the first one, which is not great (tag doesn't work at the moment).

How the logic between rules will look like ? 
Ctrl-scroll doesn't work at the moment.

AlicVB commented:
First, thanks for your test. I'll try to answer point by point...

> I like the button `revert to a previous set of rules` (avoids to open another module). It could be put in the module header or in a section of presets.

I guess that your goal is to remove entirely the last line with the buttons ? not sure about your proposal : module headers are standardized and I'm not fond of adding an exception here... same apply for preset menu about standardization, not to speak that I expect the preset menu to have lot of entries very quickly...

> The expansion button could be included in the header of the sub-modules to save space.

good idea. I'll have a try

> It should be also expanded by default.

As said before, imho this should be expanded by default for folders, tags, ... but it's way less useful for rating, colorlabels, ... How it is implemented atm, it's very easy for us to change default state : just an entry in `darktableconfig.xml.in` so it'll set the default behavior for each property independently..

> Appending a preset is great as a function, but the long list of rules is not appealing. It would more logic to append presets from the presets menu (probably having a specific section for them). Currently they appear in presets but act as normal presets which not (mandatorily) wanted. The long list could be also reduced by a configuration as for `image information` to keep only the required ones.

Clearly the long list is not appealing... I'm not fond of using the preset menu for that, because it's a core things and this would require to add an exception in the core, specifically for collect lib which should be avoided if possible. not to speak that I expect the list of preset to become quite long too very quickly...
Instead of that, I would prefer a different approach to tackle that issue (which already exist in current implementation, btw) : 
- we have @quovadit proposal with 2 buttons
- we have @EC1000 proposal about the "quick add" by typing
- we should be able to remove some entry too (in date-time part for example) by proposing a better UI
- I also take note of your "image info" like proposal
- I've some other idea, but they will need some experiment first...

> 
> If I want to change the main rule from folder to tag I have to add a new rule and remove the first one, which is not great

just click on the property and you should be able to change it (or there's a bug ;))

> (tag doesn't work at the moment).

I'll have a look

> How the logic between rules will look like ? Ctrl-scroll doesn't work at the moment.

If you means the "and/or..." they are located in the rule header, and they can be changed by clicking on them...
For ctrl-scroll, yes, I know, it seems that current implementation doesn't work well in the case of this lib. Esp. it seems to use the height of the first element to determine the total height...

phweyland commented:
Thanks for the answers.
> I guess that your goal is to remove entirely the last line with the buttons ?

No, this is not the goal. 
When one talks about presets, there is already a standard place for that. And appended presets could turn as standard as primary presets, including in darkroom. So I understand yours points but the logical place remains there.
About the last collections, please look at almost all applications: when you open a document there is the option to open the recently opened ones. Again, I understand your arguments, but the logical place for this is at the opening of the module.
This said, specific buttons work. If needed they can be moved afterwards.

>  just click on the property
> "and/or..." they are located in the rule header

Got them. That works.

> > (tag doesn't work at the moment).
> 
> I'll have a look

Should be glitch of mine, because it works now. Sorry for the noise.

AlicVB commented:
Little update with @phweyland recent comments : 
- rework completely the menu to choose the property : show directly the 6 most used properties (nb to show is configurable) and all the other in a submenu. That drastically lower the nb of items... (obviously the default property to show on the first use is configurable too)
- the expander icon is moved to the rule header (with the "tree" icon. It's a little bit less discoverable, but it gain 1 line and height is quite precious here !
- expand by default filmroll, folders and tags (again, the list of which property should be extended by default or not is to be discussed once the design will be stabilized)

Now I wonder if we shouldn't draw a specific icon for each property, that would be shown in the menu and eventually in the rule header (and also in the filter bar if we go the route to add several items there)...

elstoc commented:
Only time for a quick play this evening and I'm not sure how "finished" this all is so a few comments in no particular order:

- Overall I find the UI quite cluttered and confusing. I'm aware that's not very productive comment but I might have something more constructive when I've had time to consider it a bit more
- The module starts with "film roll" selected and quite a large scrolled area to hold the film roll list. I can't resize the panel with Ctrl+scroll (this key combination does some odd stuff) 
- When I choose to "add rule" it looks like nothing happens, because the new rule is hidden beneath the scrolled area. It would be better to collapse the film roll list before showing the new rule.
- The expand/collapse icon for multi-selection might be better as an up/down arrow? I've considered this as an option for expandable module sections as well (the triangle icon is overused IMO). 
- The separation between the text box at the top and the list of properties isn't obvious and at first, when I double-clicked on a filmroll it wasn't clear to me why the name of that film roll (with the entire path, even though I hide all but the end folder name in the UI) appeared in the text box. Perhaps if you could include some default text in the box (like the module search box in the darkroom) that might help, but clearer separation is definitely needed.
- I don't really like the placing the majority of filter properties beneath a sub-menu (I don't like it on the module group setup either) - it's often just a bit of extra faff to get to the option I want. I get the idea (to reduce the size of the menu) but there must be a better way (perhaps just grouping common options at the top of the list?). If we stick with this method, the name "other rules properties" doesn't make sense in EN - perhaps just "other properties". 
- It's not obvious that clicking on the "film roll" label is supposed to bring up a menu - I discovered by random-clicking.
- I'm not sure about the positioning of the actions at the bottom. I would prefer if we could include those actions within the "rules" and "sorting" sections, perhaps with a simple "+" icon?

As I say, I'm not sure how helpful I've been here. I hope to have some time to test/consider further soon.

AlicVB commented:
@elstoc : thanks for your (even quick) test. I'll try to answer some of your remarks : 



> * The module starts with "film roll" selected and quite a large scrolled area to hold the film roll list. I can't resize the panel with Ctrl+scroll (this key combination does some odd stuff)

yeah, that's a known issue with the current "general" implementation of the lib resizing. For now, I've reverted to a standard, fixed sized scrollwindow... until I fix that... (so now that should work better but you can't change the height with ctrl-scroll)



> * When I choose to "add rule" it looks like nothing happens, because the new rule is hidden beneath the scrolled area. It would be better to collapse the film roll list before showing the new rule.

good point. should be fixed.



> * The expand/collapse icon for multi-selection might be better as an up/down arrow? I've considered this as an option for expandable module sections as well (the triangle icon is overused IMO).
> * Maybe putting the expander button beside the text entry might be clearer too?

The expander was under the textbox at first, but as @phweyland as pointed, that's a lot of lost vertical space... using arrow, the main issue is that it will be confusing and I fear that user take it as a combox dropdown (even more if you take next point into account)



> * It's not obvious that clicking on the "film roll" label is supposed to bring up a menu - I discovered by random-clicking. We have a standard UI for comboboxes and we should use it.

that was my first intention, but the problem is that you will have the property aligned to the left (or right) and not centered, which would look very strange... I've also done some mockup of what would happen if we manage to hack combobox to center text, and that's strange too, with a drop down very far from the text (and close to the right icons)



> * I'm not sure about the positioning of the actions at the bottom. I would prefer if we could include those actions within the "rules" and "sorting" sections, perhaps with a simple "+" icon?

That would use 2 lines instead of one (not to speak that the "history" button will be orphaned...)



> * The separation between the text box at the top and the list of properties isn't obvious and at first, when I double-clicked on a filmroll it wasn't clear to me why the name of that film roll (with the entire path, even though I hide all but the end folder name in the UI) appeared in the text box. Perhaps if you could include some default text in the box (like the module search box in the darkroom) that might help, but clearer separation is definitely needed

I've just added some margin under the textbox (can be done via css.
For the default text, we have to keep in mind that the textbox is *not* a search box : it's the exact widget that is show right to the property combobox in actual implementation : it contain the value for the query. So empty value as a meaning (select all)

> * I don't really like the placing the majority of filter properties beneath a sub-menu (I don't like it on the module group setup either) - it's often just a bit of extra faff to get to the option I want. I get the idea (to reduce the size of the menu) but there must be a better way (perhaps just grouping common options at the top of the list?). If we stick with this method, the name "other rules properties" doesn't make sense in EN - perhaps just "other properties".

Text changed.
The idea here is based on the fact that most user will use only a few number of property most of the time, but that properties will change from one user to another, hence the idea to put a "direct" access to the 6 most used properties without preventing the user to use other property if he want. That has also the advantage to put most used properties at the top... (and if someone use routinely more than 6 different property, the number can be changed in darktablerc (we can also changed it by default if we have a consensus)


> * Overall I find the UI quite cluttered and confusing. I'm aware that's not very productive comment but I might have something more constructive when I've had time to consider it a bit more

To reduce cluttering, I've remove the "section" labels, and replaced it by some empty space...

That said, it seems that you are not alone, sadly :( 
That's not a problem at all, but... tbh, I'm mitigated about what to do now... In one hand I don't want to waste time in implementing an unusable UI... in other hand I find this UI not so bad (even if it needs refining and fixes) and I can't really see an completely different way to do it in order to allow : 
- have more "friendly" widgets (that may eventually take some space)
- avoid the treeview that appear far away from the "active" rule
- separate the "completion/search" feature from the list of proposals
- regroup all that "collect" features in a same place (collect ; sort; history ; (filtering to be done))
- more explicit operators / close widgets
- allow to change the orders of rules
- allow multiple sort

As always, I'm opened to all constructive proposals...

quovadit commented:
I agree with lots of the comments by elstoc:
- "Overall I find the UI quite cluttered and confusing"
  - the blocks of multiple criteria are not easy to identify, because only raw UI has a dark background (see left image)
  - it would be better if even friendly widget had a dark background (see right image)
- "It would be better to collapse the film roll list before showing the new rule."
- "The expand/collapse icon might be better as an up/down arrow"
  - I liked the arrow at the bottom of each block, to save space it could be much narrower
- "It's not obvious that clicking on the [tag] label is supposed to bring up a menu"
- "I would prefer if we could include the actions within the "rules" and "sorting" sections, perhaps with a simple "+" icon"
  - add rule --> + rule
  - add sort --> inside 'sorting' block
  - history-icon --> history-button

And to repeat some of my previous comments:
- I would definitely split appending presets from adding rules with two separate buttons
- still I would increase the space between the blocks
- sorting should be collapsed in a separate block, because it's not necessary in most cases
- I would also add a separate button to 'load' presets (other than '+ presets' which appends a preset) 

![image](https://user-images.githubusercontent.com/5677563/150017393-8f7a0e89-4457-423d-8ea9-81550eecd4ec.png)

elstoc commented:
> The expander was under the textbox at first, but as @phweyland as pointed, that's a lot of lost vertical space... using arrow, the main issue is that it will be confusing and I fear that user take it as a combox dropdown

This is why I would suggest doing away with the "triangle" icon for anything except comboboxes and coming up with an alternative for collapsible UI. It is confusing to have an icon used for multiple things. When I said "arrow" I meant like --> with a stem to differentiate.

Regarding lost vertical space, I'm talking small buttons with an icon so I'd hope they could maybe consume horizontal space somewhere instead.

> In one hand I don't want to waste time in implementing an unusable UI...

This is a horrible module to make nice UI for and getting it right will be hard work so kudos to you for grabbing the nettle. Sadly though, I think the solution from a UI point of view is probably to make it do _less_ rather than _more_. I'm all for masses of flexibility but if it makes things complicated to use it's not helping anyone.

Nilvus commented:
It's a really great work but the UI will be too cluttered. So we need to think more about how to improve that part before going on more. I think something your proposal, @AlicVB and that mockup from @Ni-Ti in #9664 will be better (just forget my critical comment in that issue, my view on that had changed since).

I will try to think more about some mockup ASAP. I have had other priorities those last weeks but I should/could (I hope) spent some time to think more on that this week-end. I ping @difrkguilar as he had some interesting mockup proposal. If he have/want to propose some ideas!

SoupyGit commented:
Sorry I can't comment on the work done, but I too like the mockup by @Ni-Ti in #9964. Looks neat, and simple to understand.

quovadit commented:
Of course the mockup in 9964 is simpler to understand, but the goal of this PR is more than only simplifying existing functionality:
- I love the possibility to switch between raw and friendly widgets
- it's great to have the possibility to append presets to existing filters
- sorting can be saved as well, and you can even define multiple sort-criteria

Regarding UI, I think it's a huge improvement to have two-row-blocks instead of the one-liner, but I still think the overall look is not clean enough yet.

For me there are two possibilities to make the UI cleaner:
- split the functionality to two or three modules: navigate / filter [ / recent]
  - see mockup in [10643](https://github.com/darktable-org/darktable/issues/10643#issuecomment-999912297)
- keep everything in one module, but collapse sorting
  - see right mockup [above](https://github.com/darktable-org/darktable/pull/10733#issuecomment-1015827611)

AlicVB commented:
Thanks for all your comments :)
@Nilvus : no problem about the delay. I perfectly know what real life can be ;)

I've seen the mockups in #9964 too. But as @quovadit says one of the main point here is to provide more friendly widgets, but that require space, way more than actual "half-panel-width" (and we don't speak about adding the operator on the line...)
in fact, I even wonder if at some point we shouldn't propose a complete "raw" interface (like a simple big textarea) and one with only friendly widgets...

I've done some improvement (I think) with latest commit on the "visibility" part : 
![Capture d’écran_2022-01-20_20-05-02](https://user-images.githubusercontent.com/1818223/150405440-2f63986c-5c82-4821-8a33-2f6cbafe8d22.png)
- put back the expander on the bottom  : that makes a nice separator and tidy up the header
- add some hover effect on the property to better indicate that click has an effect (I've tried to add some "false" arrow like @quovadit suggest, but imho that make the UI more cluttered
- accentuate the contrast to make the rules more distinct
- put friendly widgets on a distinct background

next step is to implement @quovadit proposal for the buttons...

quovadit commented:
wow, that looks perfect!

can't wait to see the new buttons...

AlicVB commented:
New version with the different buttons...
Note that the lib sizing is still not working (currently, I've entirely removed it, even the scrollwindow, to avoid strange effects)
Same for multiple sorts that are not implemented (the `+` after the sort line does nothing)

Nilvus commented:
This is coming better and better. Really great work. I will test more now but just some quick feedback:

- think about ellipsize buttons text. For example, on my panel width, I see preset like that and was afraid at start on what was that button (seeing no ...)
![image](https://user-images.githubusercontent.com/45535283/150698111-529278d6-1dcc-45b6-b62b-94a35da0b81b.png)
- I know that this will not help on width size (or height one depending way to deal with that) but I really think + sign on rule and preset and button is a bad way. We have for me here 2 sets of button:
  - add ones (rule and preset)
  - load ones (load and history)
So maybe something to do with that. I don't know maybe a title above to allow having only rule and preset. Or put add buttons at the top of the modules and load ones where they actually ares. Yes, that add lines but allow to add text and made better buttons. It's just ideas, maybe thinking more about it will made someone here propose better idea
- there's too much space between those buttons and sort by filter toolbar

About that filter toolbar, you will soon be able to change the arrow by my new sort by icon when my PR will be merged!

quovadit commented:
for me the new layout is perfect! 

It's very clean, easy to understand what's going on, and very simple to use.

I don't think that splitting the buttons to different locations would be a good idea, because I think it's easier to understand the difference between adding and replacing if both options are next to each other.

I also like the space between buttons and sorting, because these are totally different things, and this way you don't even need to collapse sorting (as in my mockup).

And meanwhile I have to say it's really great to have the history within the collect module, to quickly 'undo' a selection

TurboGit commented:
I understand that this is very hard work... Let me see if we can do even better :) What about having a way to group (at least one level) condition to be able to mix and / or conditions more easily. The parenthesis could be graphically conveyed using indentation?

`   rating=1 and color_label=red or iso>100`

is ambiguous, the could be:

`   rating=1 and (color_label=red or iso>100)`

Represented as:

```
   rating=1 
   and 
      color_label=red 
      or iso>100
```

or:

`   (rating=1 and color_label=red) or iso>100`

Represented as:

```
   rating=1 
   and  color_label=red 
   or
      iso>100
```

The idea would be to be able to add a `and` or `or` conditional for a group that will be graphically indented.

I pretty well understand that this again more complexity for the implementation. No need to do that as part of this PR, but maybe keep that in mind to be able to more easily implement that at a later time?

Nilvus commented:
@TurboGit: just a thing. This PR will be updated by @AlicVB. We talk a lot some days ago about that. This will be simplify and filter will be with a new module proposed by @AlicVB (the work is at early stages) in #11066 PR. So your comment, on what I understand is more suitable for #11066 PR.

AlicVB commented:
As @Nilvus said, we have done quite some work on this, and I'll come with a new simpler proposal soon...

To be more clear about the difference between #11066 and this one : 
- #11066 is dedicated *only* to the "friendly" widget themselves. Friendly widgets that we be used whatever the new collect/filter organization.
- this PR is dedicated to the "main" collect/filter/sort organization/disposition.

In fine both PR will be merged together, but for the discussion, it'll be simpler to not mix them together...

@TurboGit : interesting proposal... but I fear that it'll be very challenging to propose a simple UI to let user choose that, especially to let one "set" the parenthesis. At some point, I think we'll need to come with a "full raw" interface that would allow complex and advanced things like what you propose. I have in mind a "simple" multi-line text widget where one can write manually the rules, exactly in the form you have proposed.

dterrahe commented:
@AlicVB Is this completely OBE and closeable?

AlicVB commented:
yep. Forgot this one ;)
There will be further change for next releases, but this PR is completely obsolete by now...
